### ISSUE TITLE
Referencing a WinRT component generates a warning, no implementation found for winmd

### ISSUE THREAD

**Issue by jeff-r-skillrev** [NuGet/Home#1508]
I'm in a similar situation to @furmek on #1221 but my project is a WinRT component written in C++. I have a .winmd and 3 architecture specific .dll files (x86, x64, and ARM), but my .dll is NOT managed. As such, I can't use corflags to convert it to a reference implementation.

My existing solution works, but I get a warning about it not being able to find the implementation for lib\uap10.0\curve25519.winmd.

```
Severity    Code    Description Project File    Line
Warning     No implementation file was provided for the .winmd file 'C:\Users\langboost\.nuget\packages\curve25519-uwp\1.0.3\runtimes\win10-x86\lib\uap10.0\curve25519.winmd'. To generate registration information in the app manifest, specify the 'Implementation' metadata on the .winmd reference item in the project file.
```

How can I make some kind of reference DLL for a WinRT project? I want to remove this warning, and ideally see Visual Studio generate appropriate app manifest information for this NuGet package.

Here is my nuspec file:
https://github.com/langboost/curve25519-uwp/blob/master/deploy/nuget/curve25519-uwp.nuspec
I have a targets file too, but only so that the *.h files are included if the package is used by C++.

Any samples or advice would be appreciated. I'm more than happy to contribute code if you can point me to the area where NuGet makes decisions about what kind of reference is added to the project on behalf of my package.

**Comment by yishaigalatzer** [comment]
Can you clarify a bit more about your scenario.

Are you trying to install your package into a C++ project? Or are you building your package in a C++ project but expect it to be consumed by C++ and C# projects (basically all project types)

 CC @ericstj @MeniZalzman @csharpfritz - Docs for this need to go live ASAP

**Comment by jeff-r-skillrev** [comment]
In my case, I am consuming a C++ WinRT component in a C# class library. For the purposes of this issue, I'm primarily concerned with using my WinRT component from a managed library, but will eventually verify it can also be consumed from a C++ library.

-----Original Message-----

**Comment by jeff-r-skillrev** [comment]
If this is being categorized as docs, does that mean there is something I can do now to get this warning to go away? What is it?

**Comment by ericstj** [comment]
That warning means you're either missing the implementation metadata from the WinMD item or it's invalid.  It would actually result in failed activation of your component in some cases.

Just package your WinMD in `lib` since a WinMD for a native winrt component is just a metadata assembly which is architecture agnostic.

For your native DLLs put then in `runtimes/{rid}/native`.  By defining both `lib/uap10.0` and `runtimes/{rid}/lib`, you are actually superseding the RID-less winmd @ `lib/uap10.0`.

You can actually relax the error in your targets to only apply when '$(TargetExt)' == '.exe' since an AnyCPU dll will work just fine referencing your WinMD.  You just need to make sure that the App that eventually includes the package is architecture specific.

Here is a generic set of targets I put together:

```
<?xml version="1.0" encoding="utf-8"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">

  <Target Name="_ValidatePlatformFoo" BeforeTargets="Build">
    <Error Condition="'$(TargetExt)' == '.exe' AND '$(Platform)' != 'x86' AND '$(Platform)' != 'x64' AND '$(Platform)' != 'ARM'"
           Text="Unsupported platform value '$(Platform)'. The $(MSBuildThisFileName) package requires the Platform to be set to x86, x64, or ARM for executable applications." />
  </Target>

  <!-- ensure we have a reference to the C++ CRT which may not be present in C#/VB/JS applications -->
  <ItemGroup>
    <SDKReference Include="Microsoft.VCLibs, Version=14.0" />
  </ItemGroup>

  <!-- If the project.json file is missing assume a packages.config
       based project and add the NuGet v3 assets.  -->
  <ItemGroup Condition="!Exists('project.json') AND !Exists('$(MSBuildProjectName).project.json') AND '$(TargetPlatformIdentifier)' == 'UAP'">

    <PropertyGroup>
      <_platformFoo Condition="'$(Platform)' == 'Win32'">x86</_platformFoo>
      <_platformFoo Condition="'$(Platform)' != 'Win32'">$(Platform)</_platformFoo>
    </PropertyGroup>

    <Reference Include="$(MSBuildThisFileDirectory)..\..\lib\uap10.0\foo.winmd">
      <!-- Native WinMDs require implementation metadata that tells the 
           packaging tasks where to look for the native dll to generate activation
           metadata, we use just the short name then set up the native dll as a
           a copy local asset.  In nuget v3 projects this is done automatically by 
           the msbuild task -->
      <Implementation>foo.dll</Implementation>
    </Reference>
    <ReferenceCopyLocalPaths Include="$(MSBuildThisFileDirectory)..\..\runtimes\win10-$(_platformFoo)\native\foo.dll" />
  </ItemGroup>

</Project>
```

And here's the layout I'd expect would work:

```
build\native\foo.targets
lib\uap10.0\foo.winmd
runtimes\win10-arm\native\foo.dll
runtimes\win10-x86\native\foo.dll
runtimes\win10-amd64\native\foo.dll
```

I started writing some docs for this scenario [here](https://github.com/ericstj/NuGetDocs/blob/ericstj/nuget3samples/NuGet.Docs/Create/nuget3-packages-walkthrough.md) with [sample](https://github.com/ericstj/Samples/tree/ericstj/nuget3/NuGet3/1.5.dll.runtimes), but didn't get to the WinMD scenario.  It's slightly different than the managed DLL + native dependency scenario, but pretty similar.

I think there is still an outstanding issue on JavaScript apps where the JS project system doesn’t like the references that nuget adds.  We might need to either a) fix those in nuget, or b) have the targets remove them, to be replaced by the reference items added by the targets themselves.

I did publish a tool for generating source for a reference assembly, but I don't think you'll need that for this scenario.

**Comment by kraigb** [comment]
Adding a response from Eric when I queried about the sample and docs he refers to above:

Those docs were a time-boxed effort that was meant to show direction and the NuGet folks were going to pick it up and flesh it out.  The NuGet team did try to pick up a lot of that and translate it to documentation which I reviewed at the time.  I don’t think they had the bandwidth to flesh out the samples and do a proper walkthrough so they left that part out.

If you see some value in actually producing samples and walkthroughs I think that’s a great idea.  I’d be willing to help answer questions and review those docs but I don’t have the cycles to generate them.

Things have changed around this quite a bit, especially with the new msbuild packagereference support, and project.json is effectively dead, so I don’t think using my docs as a starting point is all that interesting.  Taking the high-level scenarios from them and extrapolating more is interesting.  If someone were to produce the bulk of the docs and samples and just needed some help answering questions about making sure the samples work I could help with that.  I think this would be especially interesting for someone on the NuGet team since it would help them see where the gaps are in their product firsthand.
